core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -562,8 +562,9 @@ public void refresh(String source) throws EngineException {
             ensureOpen();
             searcherManager.maybeRefreshBlocking();
         } catch (AlreadyClosedException e) {
-            ensureOpen();
```

Comment:
`ensureOpen()` can throw its own exception, masking the (unexpected) `AlreadyClosedException` we just hit.  Also, except for disaster (failing engine), we should never have hit an `AlreadyClosedException` because the engine was closed in that try block since we hold the `readLock`.  I think?